File: D star Lite C++/D_start_Lite/src/ComputeShortestPath.cpp
/* Include files */
#include <iostream>
#include <stdio.h>

#include "main.h"
#include "Initialize.h"
#include "ComputeShortestPath.h"
#include "CalculateKey.h"
#include "GetMinKey.h"
#include "CompareKey.h"
#include "isSamePosition.h"
#include "UpdateVertex.h"

/* Parameters Declarations */
static int minIndex;
static double cost;

/* Matrix Declarations */
static RowVector2d key_s_start;
static RowVector2d key_old;
static RowVector2d key_new;
static RowVector2d key_TopKey;
static RowVector3d temp_key;
static RowVectorXd u;
static RowVector2d s_temp;

/* Function Declarations */
void ComputeShortestPath(void)
{
    key_s_start = CalculateKey(s_start);
    temp_key = GetMinKey(U);
    //TopKey取前两位
    key_TopKey = temp_key.head(2).array();


    while (CompareKey(key_TopKey, key_s_start) || (rhs(s_start(0), s_start(1)) > g(s_start(0), s_start(1))))
    {
        //获取U中最小key和index
        temp_key = GetMinKey(U);
        //minIndex取第三位
        minIndex = temp_key(2);
        //当前节点u为U中最小值
        u = U.row(minIndex);

        //key_old取前两位
        key_old = temp_key.head(2).array();

        //计算该节点u新key值
        key_new = CalculateKey(u);

        //如果该节点的key值小于原先的key值
        if (CompareKey(key_old, key_new))
        {
            //更新key值
            U(minIndex, 2) = key_new(0);
            U(minIndex, 3) = key_new(1);
        }
        //局部过一致，边缘代价函数值变低，代表网格上障碍物被清除或者搜索到一条更短捷径,开始扩展节点
        else if (g(u(0), u(1)) > rhs(u(0), u(1)))
        {
            //使局部一致
            g(u(0), u(1)) = rhs(u(0), u(1));
            //删除该节点
            removeRow(U, minIndex);

            //for all s ∈ Pred(u)
            for (int i = 0; i < 8; i++)
            {
                s_temp << u(0) + neighbour(i, 0), u(1) + neighbour(i, 1);
                //越界跳过
                if (s_temp(0) < 1 || s_temp(0) > xmax || s_temp(1) < 1 || s_temp(1) > ymax)
                {
                    continue;
                }

                if (!isSamePosition(s_temp, s_goal))
                {
                    if (c(s_temp(0), s_temp(1)) != INF)
                    {
                        cost = (s_temp.head(2) - u.head(2)).norm();
                    }
                    else
                    {
                        cost = INF;
                    }
                    rhs(s_temp(0), s_temp(1)) = fmin(rhs(s_temp(0), s_temp(1)), cost + g(u(0), u(1)));
                    //std::cout << "rhs(" << s_temp(0) << "," << s_temp(1) << ") : " << rhs(s_temp(0), s_temp(1)) << std::endl;
                }
                UpdateVertex(s_temp);
            }

        }
        else
        {
            //待补充
        }

        key_s_start = CalculateKey(s_start);
        temp_key = GetMinKey(U);
        //TopKey取前两位
        key_TopKey = temp_key.head(2).array();
    }

    
}

File: D_star Lite/D_star Lite/ComputeShortestPath.m
function ComputeShortestPath()
    global s_start;
    global s_goal;
    global U;
    global neighbour;
    global g;
    global c;
    global rhs;
    global xmax;
    global ymax;
    [key_s_start(1),key_s_start(2)]=CalculateKey(s_start);
    [key_TopKey(1),key_TopKey(2),~]=GetMinKey(U);
    while(CompareKey(key_TopKey,key_s_start)||(rhs(s_start(1), s_start(2))>g(s_start(1), s_start(2))))%当优先列表中最优点（key最小）的key值小于起始点的key值或者起始点局部不一致（g和rhs不等）时

        % 此时U的首位是key值最小的元素
        [key_old(1),key_old(2),minIndex]=GetMinKey(U);
        u=U(minIndex,:);
        
        [key_new(1),key_new(2)]=CalculateKey(u);%计算该节点新key值     
        %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%         pause(0.0001)        
%         % plot current node
%         if ~isSamePosition(u, s_start)
%             red=norm(u(1:2)-s_start)/norm(s_goal-s_start);
%             if red>1
%                 red=abs(2-1*red);
%                 green=0;
%             else
%                 green=1-red;
%             end
%             node_color=[red,green,0];
%             plot(u(1),u(2),'o','MarkerFaceColor',node_color,'MarkerEdgeColor',node_color)
%         end
        %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

        if(CompareKey(key_old,key_new))%如果该节点的key值小于原先的key值
            %U.Update(u,key_new)
            U(minIndex,3:4)=key_new;
            
        elseif(g(u(1),u(2))>rhs(u(1),u(2)))%局部过一致，边缘代价函数值变低，代表网格上障碍物被清除或者搜索到一条更短捷径,开始扩展节点
            g(u(1),u(2))=rhs(u(1),u(2));%使局部一致
            U(minIndex,:)=[];%删除该节点
            
            % for all s ∈ Pred(u)
            for i=1:size(neighbour,1)
                s_temp=u(1:2)+neighbour(i,1:2);
                if s_temp(1)<1||s_temp(1)>xmax||s_temp(2)<1||s_temp(2)>ymax
                    continue
                end
%                 if isInOpenList(s_temp,U)
%                     continue
%                 end
%                 if isObstacle(s_temp,obstacle)
%                     continue
%                 end
                if ~isSamePosition(s_temp, s_goal)
                    if c(s_temp(1),s_temp(2))~=Inf
                        cost=neighbour(i,3);
                    else
                        cost=Inf;
                    end
                    rhs(s_temp(1),s_temp(2))=min(rhs(s_temp(1),s_temp(2)),cost+g(u(1),u(2)));
%                     fprintf("rhs(%d,%d): %d\n",s_temp(1),s_temp(2),rhs(s_temp(1),s_temp(2)))
                end
                UpdateVertex(s_temp);%遍历前继节点集并更新前继节点的key值
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%                 % plot temp next node
%                 plot(s_temp(1),s_temp(2),'o','color','b')
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
            end
            
%%%%%%%%%%%%%%%% REACH TO HERE AFTER NEW OBSTACLE SCANNED %%%%%%%%%%%%%%%%
        else
            g_old=g(u(1),u(2));
            g(u(1),u(2))=Inf;%遇到障碍物
            predecessor_list=[u(1:2)+neighbour(:,1:2)
                              u(1:2)];

            for j=1:size(predecessor_list,1)
                s_temp=predecessor_list(j,:);
                if s_temp(1)<1||s_temp(1)>xmax||s_temp(2)<1||s_temp(2)>ymax
                    continue
                end
%                 if c(u(1),u(2))==Inf || c(s_temp(1),s_temp(2))==Inf
%                     c_su=Inf;
%                 else
%                     c_su=norm(u(1:2)-s_temp);
%                 end
%                 if rhs(s_temp(1),s_temp(2))==c_su+g_old
                if ~isSamePosition(s_temp,s_goal)
                    rhs_cal(s_temp);
                end
%                 end
                UpdateVertex(s_temp);
            end
%%%%%%%%%%%%%%%% REACH TO HERE AFTER NEW OBSTACLE SCANNED %%%%%%%%%%%%%%%%

        end
        [key_s_start(1),key_s_start(2)]=CalculateKey(s_start);
        [key_TopKey(1),key_TopKey(2),~]=GetMinKey(U);
    end
end